### DISCUSSION TITLE
mysql Mixed queries I can't get lastrowid . (ver 1.4.18)

### DISCUSSION THREAD

**Discussion by Honjoy** [sqlalchemy/sqlalchemy#6647] [Usage Questions]
I have try get lastrowid, always return 0, but in sqlalchemy version 1.3.24 and fastapi_sqlalchemy is return the correct value.

python code:
```py
from fastapi import FastAPI
from fastapi_async_sqlalchemy import SQLAlchemyMiddleware
from fastapi_async_sqlalchemy import db

app = FastAPI()
app.add_middleware(
    SQLAlchemyMiddleware, 
    db_url="mysql+aiomysql://user:user@192.168.88.200:5432/test1"
)


@app.get("/")
async def get_last_id():
    result = await db.session.execute("update test_table set pname='test name' where id=1; insert test_table2 (pname) values ('test name val')")
    lastid=result.lastrowid
    await db.session.commit()
    return lastid

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8002)
```

**Versions.**
 - OS: debian 10
 - Python: 3.7
 - SQLAlchemy: 1.4.18
 - Database: mysql 5.7
 - DBAPI: aiomysql 0.0.21

**Comment by CaselIT** [comment]
Hi,

The issue may be the different dbapi. What dbapi did you use on 1.3 instead of aiomysql?
Does that dbapi work the same in 1.4?

**Comment by Honjoy** [comment]
version1.3 use default mysqlclient 2.0.3

**Comment by CaselIT** [comment]
Can you try `mysqlclient ` with 1.4 and the same query?

**Comment by Honjoy** [comment]
maybe is correct  but mysqlclient can not use in async mode.
If I want to get lastrowid, I can't use the async mode？

**Comment by CaselIT** [comment]
> 
> 
> maybe is correct but mysqlclient can not use in async mode.
> If I want to get lastrowid, I can't use the async mode？

I'm trying to understand if the issue is of aiomysql or in sqlalchemy. If it's in the dbapi there isn't much we can do I fear.
I'm just asking if you can try running the same query in an script to see if it works using the `mysqlclient ` with 1.4

**Comment by Honjoy** [comment]
sorry,  I can't try it in this time, sqlalchemy 1.4 can not use fastapi-sqlalchemy, and I don't know how to use sqlalchemy core for this.

**Comment by zzzeek** [comment]
hi there -

".lastrowid" does not apply to UPDATE statements.  it only applies to INSERT statements.

From https://www.python.org/dev/peps/pep-0249/:

> Cursor.lastrowid
>    This read-only attribute provides the rowid of the last modified row (most databases return a rowid only when a single INSERT operation is performed). If the operation does not set a rowid or if the database does not support rowids, this attribute should be set to None.

**Reply by zzzeek** [reply]
can confirm aiomysql returns 0 in this case.   there's no reason to get "lastrowid" for an UPDATE in any case.

```python

import asyncio

from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import MetaData
from sqlalchemy import String
from sqlalchemy import Table
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine


meta = MetaData()

t1 = Table(
    "t1", meta, Column("id", Integer, primary_key=True), Column("name", String(50))
)


async def async_main():
    # engine is an instance of AsyncEngine
    engine = create_async_engine(
        "mysql+aiomysql://scott:tiger@localhost/test",
        echo=True,
    )

    async with engine.begin() as conn:

        await conn.run_sync(meta.drop_all)
        await conn.run_sync(meta.create_all)

        # for normal statement execution, a traditional "await execute()"
        # pattern is used.
        await conn.execute(
            t1.insert(), [{"name": "some name 1"}, {"name": "some name 2"}]
        )

    async with engine.begin() as conn:
        result = await conn.execute(
            text("update t1 set name='new name'")
        )

        # actual aiomysql cursor is here, this returns "0"
        print(result.context.cursor.lastrowid)

    await engine.dispose()

asyncio.run(async_main())
```

**Reply by CaselIT** [reply]
I think the issue is that an insert is also issued in the same execute, and apparently aiimysql does not registers it

**Reply by zzzeek** [reply]
like from a trigger?    that seems like a leap to be guessing that here but in any case the user can invoke the explicit MySQL calls by calling "SELECT LAST_INSERT_ID()":

https://dev.mysql.com/doc/refman/8.0/en/information-functions.html#function_last-insert-id

that would be the better way for unusual "lastrowid" cases to be handled here as this would be highly mysql specific

**Reply by CaselIT** [reply]
The original query was ` result = await db.session.execute("update test_table set pname='test name' where id=1; insert test_table2 (pname) values ('test name val')")`

I guess that two executions can be used or the `select last_row_id` can be used.

**Reply by zzzeek** [reply]
oh, i didn't scroll it rightwards, sorry.

you *definitely* should not be bundling queries with a semicolon like that.  that is totally not supported.
